Cep 29 not operator

pkolaczk opened the pull request:
* NOT IN
* NOT CONTAINS / NOT CONTAINS KEY
* Inequality operator (!=) for map entries
* Inequality operator (!=) for clustering columns
* Both ALLOW FILTERING and indexing (SAI) supported

Commit:
CQL NOT CONTAINS, NOT CONTAINS KEY, != for map entries

Ability to use NOT operator in filters on lists, sets and maps.
Requires ALLOW FILTERING.

All those operators are implemented in a ContainsRestriction, therefore
they allow to be mixed: i.e. you can use
multiple CONTAINS, NOT CONTAINS, CONTAINS KEY, NOT CONTAINS KEY
in a single WHERE clause.

Commit:
CQL NOT IN operator

Ability to use NOT IN restrictions in CQL WHERE clause.
Can be used on any column with ALLOW FILTERING.

Can be also used on clustering columns w/o ALLOW FILTERING.
Multicolumn NOT IN restrictions are supported
(e.g. (c1, c2) NOT IN ((1, 2), (3, 4))).

NOT IN restrictions are implemented as slice restrictions, therefore
they can be merged with other slice restrictions using <, >, <=, >=.
Also multiple NOT IN restrictions can be merged together.

Overview of the code changes:
  - NOT IN operator has been added to the CQL WHERE clause parser
    definition.

  - SliceRestriction got a new field with a list of terms to skip,
    received from the terms passed in the NOT IN clause.

  - SliceRestriction methods have been adjusted to generate filters
    that filter out the items to skip, as well as generate proper
    clustering bounds, omitting the items to skip.

  - Because the list of items to skip can be given either as a marker
    or a list literal, MarkerOrList has been
    introduced to abstract out the concern of getting the items out of
    the SliceRestriction. This way both SliceRestriction and
    INRestriction could be reduced. INRestriction hierarchy has been
    flattened to a single class.

  - MultiCBuilder has been rewritten. It has been both simplified and
    added new functionality. The specialization for single column
    clusterings has been removed as it had almost doubled the
    complexity for a questionable performance gain. The code for
    computing cartesian products has been simplified and optimized:
    now it has proper O(n^2) complexity instead of O(n^3).
    MultiCBuilder now also tracks the inclusiveness of each clustering
    bound. This is required because when merging NOT IN with inequality
    like <= or >=, we need to mix bounds differing in inclusiveness.

  - MultiColumnRestriction.appendBoundTo has been modified to use
    the new MultiCBuilder and also simplified and documented, as
    this part of code turned out to be extremely hard to understand
    for me.

Commit:
Allow CQL equals / non-equals operators in more places

Now that we implemented NOT IN, both equals and non-equals
can be rewritten as slices.

column != value ---> column NOT IN (value)
column  = value ---> column >= value AND column <= value

And because slices can be freely merged, this opens room for
many interesting queries that were previously
impossible, e.g.:

SELECT ... WHERE c1 != ?
SELECT ... WHERE c1 != ? AND c1 != ? AND ...
SELECT ... WHERE c1 = ? AND (c1, c2) != (v1, v2)
SELECT ... WHERE c1 IN ? AND (c1, c2) != (v1, v2)

Commit:
SAI acceleration of NOT CONTAINS / NOT CONTAINS KEY

NOT CONTAINS, NOT CONTAINS KEY and map entry inequality
can be now used on SAI indexed columns without need for
ALLOW FILTERING.

The result sets are evaluated by taking an iterator
to all primary keys (in all active memtables and sstables)
and anti-joining them with the primary keys
matching the CONTAINS predicate, read from the index.
This way there was no need to change the index structure.
Although a full scan over all primary keys is needed,
the scan involves scanning of primary keys only, without deserializing the whole rows.
For sstables, the primary key trie component is used, which is likely much smaller than
the whole sstable. Therefore, this method is likely faster than filtering, as long
as rows are big enough.

adelapena commented on src/antlr/Parser.g:
```diff
@@ -1789,6 +1793,16 @@ relation[WhereClause.Builder clauses]
+              { $clauses.add(MultiColumnRelation.createSingleMarkerNotInRelation(ids, tupleInMarker)); }
+          | literals=tupleOfTupleLiterals /* (a, b, c) IN ((1, 2, 3), (4, 5, 6), ...) */
+              { $clauses.add(MultiColumnRelation.createNotInRelation(ids, literals)); }
+          | markers=tupleOfMarkersForTuples /* (a, b, c) IN (?, ?, ...) */
```
The three comments should use `NOT IN` instead of only `IN`

pkolaczk replied:
Fixed

adelapena committed:
Use inheritance in MarkerOrList/MarkerOrTerms

adelapena commented on src/java/org/apache/cassandra/cql3/MarkerOrList.java:
```diff
@@ -0,0 +1,140 @@
+ * "SELECT ... WHERE x IN (?, ?, ?)"
+ * "SELECT ... WHERE x IN (a, b, c)"
+ */
+public class MarkerOrList
```
Maybe this could be named `MarkerOrTerms`, since without context it's not very clear what kind of list it is. Also, the constructor and most methods are branched depending on whether it's a single marker or a list of terms. This makes me think that the class could benefit from inheritance, more or less this way: https://github.com/adelapena/cassandra/commit/1a2361db14e4f1e2c32fbda4a0a3b8dd8543ea66

pkolaczk replied:
I'm not a very big fan of inheritance / subclassing but indeed in this case it is cleaner. Thank you. I merged your commit.

adelapena commented on src/java/org/apache/cassandra/cql3/MultiColumnRelation.java:
```diff
@@ -95,6 +96,11 @@ public static MultiColumnRelation createInRelation(List<ColumnIdentifier> entiti
         return new MultiColumnRelation(entities, Operator.IN, null, inValues, null);
     }
 
+    public static MultiColumnRelation createNotInRelation(List<ColumnIdentifier> entities, List<? extends Term.MultiColumnRaw> inValues)
```
This could have JavaDoc like the method `createInRelation` right above.

pkolaczk replied:
Fixed

adelapena commented on src/java/org/apache/cassandra/cql3/Relation.java:
```diff
@@ -169,6 +214,17 @@ public final Restriction toRestriction(TableMetadata table, VariableSpecificatio
+     *
+     * @param table the table meta data
+     * @param boundNames the variables specification where to collect the bind variables
+     * @return a new EQ restriction instance.
```
```suggestion
     * @return a new NEQ restriction instance.
```

adelapena commented on src/java/org/apache/cassandra/cql3/Relation.java:
```diff
@@ -203,7 +270,7 @@ protected abstract Restriction newSliceRestriction(TableMetadata table,
      * @throws InvalidRequestException if the <code>Relation</code> is not valid
      */
-    protected abstract Restriction newContainsRestriction(TableMetadata table, VariableSpecifications boundNames, boolean isKey);
+    protected abstract Restriction newContainsRestriction(TableMetadata table, VariableSpecifications boundNames, boolean isKey, boolean isNot);
```
The new `isNot` parameter should probably be documented. However, given that EQ/NEQ and IN/NOT_INT pairs have separate `new*Restriction` methods for the "not" variant, perhaps we should do the same here and have a separate `newNotContains` method instead of the parameter?

pkolaczk replied:
I added `newNotContainsRestriction` and removed the `isNot` param.

adelapena commented on src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java:
```diff
@@ -92,61 +92,46 @@ public ClusteringColumnRestrictions mergeWith(Restriction restriction, @Nullable
                 Guardrails.inSelectCartesianProduct.guard(builder.buildSize(), "clustering key", false, state);
 
-            if (builder.hasMissingElements())
+            if (builder.buildSize() == 0)
```
There are four calls to `builder.buildSize() == 0`, maybe we could add an `MultiCBuilder#isEmpty()` method?

pkolaczk replied:
Added

adelapena commented on src/java/org/apache/cassandra/cql3/restrictions/ClusteringColumnRestrictions.java:
```diff
@@ -92,61 +92,46 @@ public ClusteringColumnRestrictions mergeWith(Restriction restriction, @Nullable
         {
             r.appendTo(builder, options);
 
             if (hasIN() && Guardrails.inSelectCartesianProduct.enabled(state))
```
I understand that the guardrails `partitionKeysInSelect` and `inSelectCartesianProduct` shouldn't consider the new `NOT IN` restrictions. That operator uses filtering and as such doesn't have the same performance issues the guardrails try to protect us from.

pkolaczk replied:
`NOT IN` creates slice restrictions not `IN` resrictions. Slice restrictions on different columns cannot be combined with a cartesian multiplication. There is only one slice restriction allowed, and the slice restriction must be present on the last restricted clustering column. 

I.e you cannot do this:
`SELECT * FROM x WHERE clustering_col_1 NOT IN (...) AND clustering_column_2 NOT IN (...)`

`NOT IN` behaves a lot more like `<, > etc` in this regard. 

However, now I noticed that `IN` can be combined with one `NOT IN`, so `NOT IN` could blow up the build size here because `NOT IN` can specify multiple values. But It looks like it is already covered - the `hasIn` check would be true in that case.

adelapena commented on src/java/org/apache/cassandra/cql3/restrictions/SingleColumnRestriction.java:
```diff
@@ -414,38 +398,56 @@ protected boolean isSupportedBy(Index index)
+        private List<Term> entryKeys = new ArrayList<>(); // for map[key] = value
         private final List<Term> entryValues = new ArrayList<>(); // for map[key] = value
+        private List<Term> negativeEntryKeys = new ArrayList<>(); // for map[key] != value
+        private List<Term> negativeEntryValues = new ArrayList<>(); // for map[key] != value
```
All can be `final`.

pkolaczk replied:
Indeed. Applied.

adelapena commented on src/java/org/apache/cassandra/cql3/restrictions/SingleColumnRestriction.java:
```diff
@@ -414,38 +398,56 @@ protected boolean isSupportedBy(Index index)
         private final List<Term> values = new ArrayList<>(); // for CONTAINS
-        private final List<Term> keys = new ArrayList<>(); // for CONTAINS_KEY
-        private final List<Term> entryKeys = new ArrayList<>(); // for map[key] = value
+        private final List<Term> negativeValues = new ArrayList<>(); // for NOT_CONTAINS
```
Nit: can we move the private constructor at the bottom of the class to the top, so it's together with the other constructors? That way I think it would be easier to understand that there can be instances with both positive and negative term lists (from `doMergeWith`).

pkolaczk replied:
Applied.

adelapena commented on src/java/org/apache/cassandra/db/MultiCBuilder.java:
```diff
@@ -19,104 +19,155 @@
+     * Either a point (single value) or a bound.
+     * Knowing the kind of elements greatly simplifies the process of building clustering bounds.
+     */
+    public static class Element
```
Nice utility class! It's not related to the patch but, since we are here, I wonder if we could/should rename the class to something more explicit, like `MultiClusteringBuilder`.

pkolaczk replied:
I agree. I like `MultiClusteringBuilder` name much more than `MultiCBuilder`.

pkolaczk replied:
Done

adelapena commented on src/java/org/apache/cassandra/db/MultiCBuilder.java:
```diff
@@ -19,104 +19,155 @@
+
+        private Element(List<ByteBuffer> values, Kind kind)
+        {
+            this.value = values;
```
Nit: the attribute and the parameter should probably have the same name

pkolaczk replied:
Fixed. Renamed field `value` to `values`. I also renamed `Element` to `ClusteringElements` as it can contain values for more than one column.

adelapena commented on src/java/org/apache/cassandra/db/MultiCBuilder.java:
```diff
@@ -19,104 +19,155 @@
-        return forMultipleValues
-             ? new MultiClusteringBuilder(comparator)
-             : new OneClusteringBuilder(comparator);
+        return new MultiCBuilder(comparator);
```
Nit: double blank line after this method

pkolaczk replied:
fixed

adelapena commented on src/java/org/apache/cassandra/db/MultiCBuilder.java:
```diff
@@ -126,15 +177,9 @@ protected void checkUpdateable()
     public int remainingCount()
     {
-        return comparator.size() - size;
+        return comparator.size() - columns.size();
```
Nit: double blank line after this method

pkolaczk replied:
fixed

adelapena commented on src/java/org/apache/cassandra/db/MultiCBuilder.java:
```diff
@@ -157,358 +202,159 @@ public boolean containsUnset()
+    public boolean hasRemaining()
+    {
+        return remainingCount() > 0;
+    }
```
Nit: double blank line after this method

pkolaczk replied:
fixed

adelapena commented on src/java/org/apache/cassandra/index/sai/disk/EmptyIndex.java:
```diff
@@ -0,0 +1,109 @@
+import org.apache.cassandra.index.sai.SSTableContext;
+import org.apache.cassandra.index.sai.iterators.KeyRangeIterator;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.utils.ByteBufferUtil;
```
Nit: unused import. Running `ant check` shows nine other unused imports, I guess CI hasn't been run yet for this patch?

pkolaczk replied:
Fixed all unused imports. `ant check` passes now.

adelapena commented on test/unit/org/apache/cassandra/index/sai/cql/CollectionIndexingTest.java:
```diff
@@ -188,6 +253,7 @@ private void createPopulatedMap(String createIndex)
             put(1, "v1");
             put(2, "v3");
         }});
+
```
Nit: unneeded blank line

pkolaczk replied:
fixed

adelapena commented on src/java/org/apache/cassandra/index/sai/IndexContext.java:
```diff
@@ -180,6 +181,11 @@ public boolean isNonFrozenCollection()
         return TypeUtil.isNonFrozenCollection(columnMetadata.type);
     }
 
+    public boolean isCollection()
```
This new method seems unused

pkolaczk replied:
Removed

adelapena commented on src/java/org/apache/cassandra/index/sai/disk/IndexSearchResultIterator.java:
```diff
@@ -61,6 +62,27 @@ public static IndexSearchResultIterator build(Expression expression,
+    {
+        KeyRangeIterator keyIterator = buildKeyIterator(expression, sstableIndexes, keyRange, queryContext);
+
+        // For NOT CONTAINS or NOT CONTAINS KEY it is not enought to just return the primary keys
```
```suggestion
        // For NOT CONTAINS or NOT CONTAINS KEY it is not enough to just return the primary keys
```

adelapena commented on src/java/org/apache/cassandra/index/sai/disk/PrimaryKeyMapIterator.java:
```diff
@@ -0,0 +1,91 @@
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+
+/**
+ * Iterates keys in the PrimaryKeyMap of an SSTable.
```
```suggestion
 * Iterates keys in the {@link PrimaryKeyMap} of a SSTable.
```

adelapena commented on src/java/org/apache/cassandra/index/sai/disk/SSTableIndex.java:
```diff
@@ -36,7 +37,9 @@
 import org.apache.cassandra.index.sai.SSTableContext;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
 import org.apache.cassandra.index.sai.disk.format.Version;
+import org.apache.cassandra.index.sai.disk.v1.V1OnDiskFormat;
```
Nit: unused imports breaking CI (`ant check`).

adelapena commented on src/java/org/apache/cassandra/index/sai/disk/v1/PerColumnIndexFiles.java:
```diff
@@ -22,6 +22,7 @@
 import java.util.EnumMap;
 import java.util.Map;
 
+import jnr.ffi.annotations.In;
```
Nit: Unneeded change

adelapena commented on src/java/org/apache/cassandra/index/sai/disk/v1/RowAwarePrimaryKeyMap.java:
```diff
@@ -161,6 +161,11 @@ public long rowIdFromPrimaryKey(PrimaryKey key)
         return triePrefixSearcher.prefixSearch(key.asComparableBytes(ByteComparable.Version.OSS50));
     }
 
+    public long count()
```
Nit: add `@Override`

pkolaczk replied:
fixed

adelapena commented on src/java/org/apache/cassandra/index/sai/disk/v1/V1SSTableIndex.java:
```diff
@@ -33,7 +34,10 @@
+import org.apache.cassandra.index.sai.disk.PrimaryKeyMap;
+import org.apache.cassandra.index.sai.disk.PrimaryKeyMapIterator;
 import org.apache.cassandra.index.sai.disk.SSTableIndex;
+import org.apache.cassandra.index.sai.disk.format.IndexDescriptor;
```
Nit: unused imports

adelapena commented on src/java/org/apache/cassandra/index/sai/iterators/KeyRangeAntiJoinIterator.java:
```diff
@@ -0,0 +1,107 @@
+        return new KeyRangeAntiJoinIterator(statistics, left, right);
+    }
+
+    protected void performSkipTo(PrimaryKey nextKey)
```
Nit: add `@Override`

adelapena commented on src/java/org/apache/cassandra/index/sai/iterators/KeyRangeAntiJoinIterator.java:
```diff
@@ -0,0 +1,107 @@
+        right.performSkipTo(nextKey);
+    }
+
+    public void close() throws IOException
```
Nit: add `@Override`

pkolaczk replied:
fixed

adelapena commented on src/java/org/apache/cassandra/index/sai/iterators/KeyRangeAntiJoinIterator.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+
+    protected PrimaryKey computeNext()
```
Nit: add `@Override`

adelapena commented on src/java/org/apache/cassandra/index/sai/iterators/KeyRangeAntiJoinIterator.java:
```diff
@@ -0,0 +1,107 @@
+        return key != null ? key : endOfData();
+    }
+
+    private int compare(PrimaryKey key1, PrimaryKey key2)
```
Nit: can be `static`

adelapena commented on src/java/org/apache/cassandra/index/sai/iterators/KeyRangeAntiJoinIterator.java:
```diff
@@ -0,0 +1,107 @@
+public class KeyRangeAntiJoinIterator extends KeyRangeIterator
+{
+    final KeyRangeIterator left;
+    final KeyRangeIterator right;
```
Nit: Can be `private final`

adelapena commented on src/java/org/apache/cassandra/index/sai/iterators/KeyRangeIterator.java:
```diff
@@ -131,6 +131,11 @@ protected final boolean tryToComputeNext()
         return hasNext;
     }
 
+    public PrimaryKey nextOrNull()
```
Nit: add `@Nullable`

adelapena commented on src/java/org/apache/cassandra/index/sai/memory/MemtableIndex.java:
```diff
@@ -30,18 +30,23 @@
 import org.apache.cassandra.index.sai.utils.PrimaryKeys;
 import org.apache.cassandra.index.sai.iterators.KeyRangeIterator;
+import org.apache.cassandra.index.sai.utils.TypeUtil;
+import org.apache.cassandra.utils.ByteBufferUtil;
```
Nit: unused imports

adelapena commented on src/java/org/apache/cassandra/index/sai/memory/MemtableIndex.java:
```diff
@@ -30,18 +30,23 @@
 {
     private final TrieMemoryIndex index;
+
+    private final IndexContext indexContext;
```
Unused attribute

adelapena commented on src/java/org/apache/cassandra/index/sai/memory/MemtableKeyRangeIterator.java:
```diff
@@ -0,0 +1,142 @@
+    protected void performSkipTo(PrimaryKey nextKey)
+    {
+        AbstractBounds<PartitionPosition> keyRange = AbstractBounds.bounds(nextKey.partitionKey(),
+                                                                true,
```
Nit: misaligned argument

pkolaczk replied:
lol, the inlay hint from Idea fooled me on this one. I stared at the code for good 10 seconds thinking wtf, they *are* aligned, and then noticed the hint that was present only at this line but not others ;)

adelapena commented on src/java/org/apache/cassandra/index/sai/memory/MemtableKeyRangeIterator.java:
```diff
@@ -0,0 +1,142 @@
+                                                                           this.keyRange.right,
+                                                                           this.keyRange.inclusiveRight());
+        DataRange dataRange = new DataRange(keyRange, new ClusteringIndexSliceFilter(Slices.ALL, false));
+        this.partitionIterator = memtable.partitionIterator(columns, dataRange, null);
```
```suggestion
       partitionIterator = memtable.partitionIterator(columns, dataRange, null);
```

adelapena commented on src/java/org/apache/cassandra/index/sai/memory/MemtableKeyRangeIterator.java:
```diff
@@ -0,0 +1,142 @@
+            {
+                Slice slice = Slice.make(nextKey.clustering(), Clustering.EMPTY);
+                Slices slices = Slices.with(this.memtable.metadata().comparator, slice);
+                this.rowIterator = memtable.rowIterator(nextKey.partitionKey(), slices, this.columns, false, null);
```
```suggestion
                Slices slices = Slices.with(memtable.metadata().comparator, slice);
                rowIterator = memtable.rowIterator(nextKey.partitionKey(), slices, columns, false, null);
```

adelapena commented on src/java/org/apache/cassandra/index/sai/memory/MemtableKeyRangeIterator.java:
```diff
@@ -0,0 +1,142 @@
+    private UnfilteredPartitionIterator partitionIterator;
+    private UnfilteredRowIterator rowIterator;
+
+
```
```suggestion

```

adelapena commented on src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java:
```diff
@@ -80,6 +80,8 @@ public TrieMemoryIndex(IndexContext indexContext)
         this.validator = indexContext.getValidator();
         this.isLiteral = TypeUtil.isLiteral(validator);
+
+
```
Unneeded change

pkolaczk replied:
fixed

adelapena commented on src/java/org/apache/cassandra/index/sai/plan/Operation.java:
```diff
@@ -41,7 +41,8 @@ public class Operation
     {
-        AND((a, b) -> a & b);
+        AND((a, b) -> a & b),
+        AND_NOT((a, b) -> a & !b);
```
`AND_NOT` is not used anywhere.

pkolaczk replied:
Yeah, it turned out not needed. I'll remove.

adelapena commented on src/java/org/apache/cassandra/utils/btree/BTree.java:
```diff
@@ -347,8 +347,8 @@ public static <Compare, Existing extends Compare, Insert extends Compare> Object
         {
-            if (isSimple(updateF))
-                return insert; // if update is empty and updateF is trivial, return our new input
+//            if (isSimple(updateF))
+//                return insert; // if update is empty and updateF is trivial, return our new input
```
Can this be removed?

pkolaczk replied:
Quite likely not. I just accidentally commeted it out during testing. Good catch!

adelapena commented on src/java/org/apache/cassandra/cql3/restrictions/TermSlice.java:
```diff
@@ -28,6 +28,12 @@
 
 final class TermSlice
 {
+    /**
```
`TermSlice#isSupportedBy(ColumnMetadata, Index)` seems to reject `NEQ` expressions. Thus, if we have:
```
CREATE TABLE t (k int PRIMARY KEY, v int);
CREATE INDEX ON t(v) USING 'sai';
```
The query `SELECT * FROM t WHERE v!=0;` will require `ALLOW FILTERING`, whereas for example the query `SELECT * FROM t WHERE v>0;` will be accepted without AF. I think both queries should be accepted without AF, shouldn't they?

We should probably include the new operators in `AllowFilteringTest`.

pkolaczk replied:
Indeed, both should be accepted.

adelapena replied:
It seems that `SELECT * FROM t WHERE v!=0` still requires `ALLOW FILTERING` after the last changes. The new check on `AllowFilteringTest#testAllowFilteringOnFirstClusteringKeyColumn` makes the test fail.

Also, `AllowFilteringTest` has test methods for different schemas. We should probably add the `NEQ` check to all the test methods, not only in `testAllowFilteringOnFirstClusteringKeyColumn`.

pkolaczk replied:
Indeed, both should be accepted. Fixed.

pkolaczk committed:
Fix comment in Parser.g (IN -> NOT IN)

pkolaczk committed:
Fix Javadoc in Relation

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Add missing JavaDoc in MultiColumnRelation.java

pkolaczk committed:
Split newContainsRestriction and remove isNot param

pkolaczk committed:
Replace buildSize() == 0 with buildIsEmpty()

pkolaczk committed:
Add missing final and move the constructor up

pkolaczk committed:
Rename CBuilder to ClusteringBuilder and add docs

Renamed:
  CBuilder -> ClusteringBuilder
  MultiCBuilder -> MultiClusteringBuilder

Added a longer explanation on how MultiClusteringBuilder works.

pkolaczk committed:
Remove double blank lines in MultiClusteringBuilder

pkolaczk committed:
Remove unused imports

pkolaczk committed:
Remove unused IndexContext#isCollection

pkolaczk committed:
Fix typo in a comment

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Format Javadoc link

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Remove unneeded `this`

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Remove uneeded blank line

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Remove extra blank line in CollectionIndexingTest

pkolaczk committed:
Remove extra blank lines in TrieMemoryIndex

pkolaczk committed:
Remove unused AND_NOT operator

pkolaczk committed:
Uncomment accidentally commented out code in BTree

pkolaczk committed:
Add final to BOTTOM, TOP and ROOT in MultiClusteringBuilder

pkolaczk committed:
Perform anti-join per each index separately to avoid false negatives

When computing non-equality searches (NEQ, NOT CONTAINS, NOT CONTAINS
KEY), perform anti-join operation on data of each index separately,
before computing the union. This may cause returning false positives
from the search, but those will be filtered out by post-filtering.

Unfortunately, each index search may return false positive matches,
due to presence of updates and deletes that happened after a flush.
Hence, we cannot compute a set difference of keys returned from one
index with the keys of another index, as subtracting a false positive
would lead to a false negative, i.e. missing keys in the
search results. We have no way to correct false negatives.

adelapena commented on src/java/org/apache/cassandra/db/MultiClusteringBuilder.java:
```diff
@@ -0,0 +1,437 @@
+import org.apache.cassandra.utils.btree.BTreeSet;
+
+/**
+ * Builder that allows to build multiple {@link Clustering}/{@link ClusteringBound} at the same time.
```
Nice doc!

pkolaczk replied:
Thank you, I hope it will make it easier for developers coming after me. I remember it took me a while until I understood what this component is for.

adelapena commented on src/java/org/apache/cassandra/db/MultiClusteringBuilder.java:
```diff
@@ -0,0 +1,437 @@
+ * </pre>
+ *
+ * Finally, we add the slice of column {@code c} by specifying the lower and upper bound
+ * (we use {@code TOP} for the upper bound), and we get the final set of clustering bounds:
```
```suggestion
 * (we use {@link ClusteringElements#TOP} for the upper bound), and we get the final set of clustering bounds:
```
Shouldn't we include `TOP` in the inclusive ends below?

adelapena commented on src/java/org/apache/cassandra/cql3/restrictions/SingleRestriction.java:
```diff
@@ -19,7 +19,7 @@
 import org.apache.cassandra.cql3.QueryOptions;
 import org.apache.cassandra.cql3.statements.Bound;
-import org.apache.cassandra.db.MultiCBuilder;
+import org.apache.cassandra.db.MultiClusteringBuilder;
```
Nit: The JavaDoc of some methods on this class still has some mentions of `<code>MultiCBuilder</code>` that should be updated to `<code>MultiClusteringBuilder</code>`. Or, even better, `{@link MultiClusteringBuilder}`.

pkolaczk replied:
Renamed everywhere and created links.

adelapena commented on src/java/org/apache/cassandra/db/MultiClusteringBuilder.java:
```diff
@@ -0,0 +1,437 @@
+    }
+
+    /**
+     * Creates a new empty {@code MultiCBuilder}.
```
```suggestion
     * Creates a new empty {@code MultiClusteringBuilder}.
```

adelapena replied:
I think this nit hasn't actually been resolved, it still says `{@code MultiCBuilder}`.

adelapena commented on src/java/org/apache/cassandra/db/MultiClusteringBuilder.java:
```diff
@@ -0,0 +1,437 @@
+ * incl end:   (a2, b3)
+ * </pre>
+ */
+public class MultiClusteringBuilder
```
Nit: There are a couple of mentions of `MultiCBuilder` in `InsertInvalidateSizedRecordsTest`.

pkolaczk replied:
Fixed

adelapena commented on src/java/org/apache/cassandra/cql3/MultiColumnRelation.java:
```diff
@@ -95,6 +96,17 @@ public static MultiColumnRelation createInRelation(List<ColumnIdentifier> entiti
 
+    /**
+     * Creates a multi-column NOT IN relation with a list of NOT IN values or markers.
+     * For example: "SELECT ... WHERE (a, b) NOT IN ((0, 1), (2, 3))"
```
```suggestion
     * For example: {code SELECT ... WHERE (a, b) NOT IN ((0, 1), (2, 3))}
     *
```

adelapena commented on src/java/org/apache/cassandra/cql3/MultiColumnRelation.java:
```diff
@@ -107,6 +119,18 @@ public static MultiColumnRelation createSingleMarkerInRelation(List<ColumnIdenti
 
+    /**
+     * Creates a multi-column NOT IN relation with a marker for the NOT IN values.
+     * For example: "SELECT ... WHERE (a, b) NOT IN ?"
```
```suggestion
     * For example: {@code SELECT ... WHERE (a, b) NOT IN ?}
     *
```

pkolaczk committed:
Allow to use SAI index for inequality search

Supports queries of the following type:
SELECT * FROM t WHERE indexed_column != ?;

pkolaczk committed:
Fix Javadoc link in MultiColumnRelation.java

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Update JavaDoc link in MultiClusteringBuilder.java

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Rename MultiCBuilder to MultiClusteringBuilder

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Update JavaDoc link in MultiColumnRelation.java

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Rename MultiCBuilder to MultiClusteringBuilder and make JavaDoc links

adelapena commented on src/java/org/apache/cassandra/cql3/restrictions/SingleColumnRestriction.java:
```diff
@@ -403,49 +381,80 @@ public void addToRowFilter(RowFilter filter, IndexRegistry indexRegistry, QueryO
         protected boolean isSupportedBy(Index index)
         {
-            return slice.isSupportedBy(columnDef, index);
+            boolean supportsSlice = slice.isSupportedBy(columnDef, index);
+            boolean supportsNeq = index.supportsExpression(columnDef, Operator.NEQ);
+            return  supportsSlice || !skippedValues.isEmpty() && supportsNeq;
```
The implementations of `SingleColumnRestriction.SliceRestriction#isSupportedBy` and `MultiColumnRestriction.SliceRestriction#isSupportedBy` are identical. Also, both implementations are the only callers of `TermSlice#isSupportedBy(ColumnMetadata, Index)`. 

Maybe we could pass the list of skipped values to `TermSlice#isSupportedBy` and delegate the NEQ check to it, putting similar things together and preventing some duplication:
```java
public boolean isSupportedBy(ColumnMetadata column, Index index, List<MarkerOrTerms> skippedValues)
{
    boolean supported = false;

    if (hasBound(Bound.START))
        supported |= isInclusive(Bound.START) ? index.supportsExpression(column, Operator.GTE)
                : index.supportsExpression(column, Operator.GT);
    if (hasBound(Bound.END))
        supported |= isInclusive(Bound.END) ? index.supportsExpression(column, Operator.LTE)
                : index.supportsExpression(column, Operator.LT);

    supported |= !skippedValues.isEmpty() && index.supportsExpression(column, Operator.NEQ);

    return supported;
}
```

pkolaczk replied:
I don't agree on this one. This is just coincidental duplication - accidentally both single and multi restrictions work the same way here, but there is no fundamental reason they have to. The amount of duplication is also very tiny. And introducing an unrelated parameter to `TermSlice` is mixing responsibilities which is IMHO worse.

I thought that maybe we could move the skipped values to the `TermSlice` itself. Then it would make more sense to have `TermSlice` handle the `isSupportedBy` logic as it would have all the information it needs. WDYT?

adelapena replied:
Agree, moving the skipped values to the `TermSlice` sounds good to me.

adelapena commented:
@pkolaczk there is a bunch of unresolved nits above that only show up if you click on the `12 hidden conversations Load more...` link on the GH web UI, maybe you have missed them?

adelapena commented on src/java/org/apache/cassandra/cql3/MultiColumnRelation.java:
```diff
@@ -95,6 +96,18 @@ public static MultiColumnRelation createInRelation(List<ColumnIdentifier> entiti
+     * @param entities the columns on the LHS of the relation
+     * @param inValues a list of Tuples.Literal instances or a Tuples.Raw markers
+     * @return a new <code>MultiColumnRelation</code> instance
+     */public static MultiColumnRelation createNotInRelation(List<ColumnIdentifier> entities, List<? extends Term.MultiColumnRaw> inValues)
```
Nit: missing line break
```suggestion
     */
     public static MultiColumnRelation createNotInRelation(List<ColumnIdentifier> entities, List<? extends Term.MultiColumnRaw> inValues)
```

adelapena commented:
I have noticed that this test returns more rows than expected:
```java
@Test
public void notInTest()
{
    createTable("CREATE TABLE %s (k int PRIMARY KEY, v1 int, v2 int)");
    createIndex("CREATE INDEX ON %s(v1) USING 'sai'");
    createIndex("CREATE INDEX ON %s(v2) USING 'sai'");

    execute("INSERT INTO %s (k, v1, v2) VALUES (1, 1, 4)");
    execute("INSERT INTO %s (k, v1, v2) VALUES (2, 2, 3)");
    execute("INSERT INTO %s (k, v1, v2) VALUES (3, 3, 2)");
    execute("INSERT INTO %s (k, v1, v2) VALUES (4, 4, 1)");

    waitForTableIndexesQueryable(currentTable());

    assertRowsIgnoringOrder(execute("SELECT * FROM %s"),
            row(1, 1, 4),
            row(2, 2, 3),
            row(3, 3, 2),
            row(4, 4, 1));

    assertRowsIgnoringOrder(execute("SELECT * FROM %s WHERE v1 NOT IN (1)"),
            row(2, 2, 3),
            row(3, 3, 2),
            row(4, 4, 1));

    assertRowsIgnoringOrder(execute("SELECT * FROM %s WHERE v1 NOT IN (4)"),
            row(1, 1, 4),
            row(2, 2, 3),
            row(3, 3, 2));

    assertRowsIgnoringOrder(execute("SELECT * FROM %s WHERE v1 NOT IN (1, 4)"),
            row(2, 2, 3),
            row(3, 3, 2)); // Got 1 extra row(s) in result: (k=1, v1=1, v2=4)

    assertRowsIgnoringOrder(execute("SELECT * FROM %s WHERE v1 NOT IN (1, 4) AND v2 NOT IN (3, 2)")); // 2 extra rows
}
```
However, the test passes if we don't create the indexes and use `ALLOW FILTERING` instead:
```java
@Test
public void notInTest()
{
    createTable("CREATE TABLE %s (k int PRIMARY KEY, v1 int, v2 int)");

    execute("INSERT INTO %s (k, v1, v2) VALUES (1, 1, 4)");
    execute("INSERT INTO %s (k, v1, v2) VALUES (2, 2, 3)");
    execute("INSERT INTO %s (k, v1, v2) VALUES (3, 3, 2)");
    execute("INSERT INTO %s (k, v1, v2) VALUES (4, 4, 1)");

    assertRowsIgnoringOrder(execute("SELECT * FROM %s"),
            row(1, 1, 4),
            row(2, 2, 3),
            row(3, 3, 2),
            row(4, 4, 1));

    assertRowsIgnoringOrder(execute("SELECT * FROM %s WHERE v1 NOT IN (1) ALLOW FILTERING"),
            row(2, 2, 3),
            row(3, 3, 2),
            row(4, 4, 1));

    assertRowsIgnoringOrder(execute("SELECT * FROM %s WHERE v1 NOT IN (4) ALLOW FILTERING"),
            row(1, 1, 4),
            row(2, 2, 3),
            row(3, 3, 2));

    assertRowsIgnoringOrder(execute("SELECT * FROM %s WHERE v1 NOT IN (1, 4) ALLOW FILTERING"),
            row(2, 2, 3),
            row(3, 3, 2));

    assertRowsIgnoringOrder(execute("SELECT * FROM %s WHERE v1 NOT IN (1, 4) AND v2 NOT IN (3, 2) ALLOW FILTERING"));
}
```

pkolaczk commented:
> there is a bunch of unresolved nits above that only show up if you click on the 12 hidden conversations Load more... link on the GH web UI, maybe you have missed them?

lol, indeed, I missed them.

pkolaczk committed:
Add missing line break in MultiColumnRelation.java

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Add missing @Override

pkolaczk committed:
Remove qualified this in MemtableKeyRangeIterator.java

Co-authored-by: Andrés de la Peña <adelapena@users.noreply.github.com>

pkolaczk committed:
Minor code style fixes (alignment, missing final/static, unused stuff)

pkolaczk committed:
Revert "Allow to use SAI index for inequality search"

This reverts commit f574547f2df21ecf06d0efc8880fccb6f4f20a71.

pkolaczk commented:
I reverted the support for `!=` and `NOT IN` on indexed columns for now as current index search does not allow multiple exclusions.

adelapena commented:
Albeit the reversal of the support for `!=` and `NOT IN` on indexed columns has made it necessary to use `ALLOW FILTERING` again, it seems that the indexes are still involved in the query and producing wrong results:
```java
@Test
public void notInTest()
{
    createTable("CREATE TABLE %s (k int PRIMARY KEY, v int)");

    execute("INSERT INTO %s (k, v) VALUES (1, 1)");
    execute("INSERT INTO %s (k, v) VALUES (2, 2)");
    execute("INSERT INTO %s (k, v) VALUES (3, 3)");
    execute("INSERT INTO %s (k, v) VALUES (4, 4)");

    String query = "SELECT * FROM %s WHERE v NOT IN (1, 4) ALLOW FILTERING";
    Object[][] rows = rows(row(2, 2), row(3, 3));

    assertRowsIgnoringOrder(execute(query), rows);

    createIndex("CREATE INDEX ON %s(v) USING 'sai'");
    waitForTableIndexesQueryable();

    assertRowsIgnoringOrder(execute(query), rows); // Got 1 extra row(s) in result: (k=1, v1=1)
}
```
I suggest we add some additional tests around the use of unsupported operators with SAI indexes and `ALLOW FILTERING`.

Mmuzaf commented on src/java/org/apache/cassandra/db/CBuilder.java:
```diff
@@ -28,9 +28,9 @@
 /**
  * Allows to build ClusteringPrefixes, either Clustering or ClusteringBound.
  */
-public abstract class CBuilder
```
Hi, I prefer to make these sorts of changes in a dedicated issue, so it looks like we can significantly reduce the number of changed files if we revert it, and keep only necessary changes here.

See the discussion below, we can ignore code formatting changes by listing the relevant commit hashes in a special file (this is not used by the project right now). So if we do this change in another issue and put that commit hash in a special file, it won't pollute the git blame in our IDEs.
https://github.com/orgs/community/discussions/5033
https://git-scm.com/docs/git-blame#Documentation/git-blame.txt-blameignoreRevsFile

This is my humble argument for such not making such a change here :-)

Mmuzaf commented on src/java/org/apache/cassandra/index/sai/disk/PrimaryKeyMapIterator.java:
```diff
@@ -0,0 +1,91 @@
+ * <p>
+ * The keys are returned in token-clustering order.
+ */
+public final class PrimaryKeyMapIterator extends KeyRangeIterator
```
The iterator looks very similar to this one, do we still need to keep it?

```
    private SSTableRowIdKeyRangeIterator(PrimaryKey min,
                                         PrimaryKey max,
                                         long count,
                                         PrimaryKeyMap primaryKeyMap,
                                         QueryContext queryContext,
                                         PostingList postingList)
```
